x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/rescorer/InferenceRescorer.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.inference.rescorer;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.Explanation;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.ScoreDoc;
+import org.apache.lucene.search.TopDocs;
+import org.apache.lucene.util.ArrayUtil;
+import org.elasticsearch.common.util.Maps;
+import org.elasticsearch.search.lookup.SearchLookup;
+import org.elasticsearch.search.lookup.Source;
+import org.elasticsearch.search.rescore.RescoreContext;
+import org.elasticsearch.search.rescore.Rescorer;
+import org.elasticsearch.xpack.core.ml.inference.trainedmodel.RegressionConfigUpdate;
+import org.elasticsearch.xpack.ml.inference.loadingservice.LocalModel;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static java.util.stream.Collectors.toUnmodifiableSet;
+
+public class InferenceRescorer implements Rescorer {
+
+    public static final InferenceRescorer INSTANCE = new InferenceRescorer();
+    private static final Logger logger = LogManager.getLogger(InferenceRescorer.class);
+
+    private static final Comparator<ScoreDoc> SCORE_DOC_COMPARATOR = (o1, o2) -> {
+        int cmp = Float.compare(o2.score, o1.score);
+        return cmp == 0 ? Integer.compare(o1.doc, o2.doc) : cmp;
+    };
+
+    private InferenceRescorer() {
+
+    }
+
+    @Override
+    public TopDocs rescore(TopDocs topDocs, IndexSearcher searcher, RescoreContext rescoreContext) throws IOException {
+        InferenceRescorerContext ltrRescoreContext = (InferenceRescorerContext) rescoreContext;
+
+        LocalModel definition = ltrRescoreContext.inferenceDefinition;
+
+        // First take top slice of incoming docs, to be rescored:
+        TopDocs topNFirstPass = topN(topDocs, rescoreContext.getWindowSize());
+        // Save doc IDs for which rescoring was applied to be used in score explanation
+        Set<Integer> topNDocIDs = Arrays.stream(topNFirstPass.scoreDocs).map(scoreDoc -> scoreDoc.doc).collect(toUnmodifiableSet());
+        rescoreContext.setRescoredDocs(topNDocIDs);
+        ScoreDoc[] hits = topNFirstPass.scoreDocs.clone();
```

Comment:
This does not do a deep copy of objects, just a copy of an array. I see that later on Line 103, we modify scores of the original hits. It is fine if we don't want to use original scores, but it looks like we still want to keep original scores?

Reply:
> It is fine if we don't want to use original scores, but it looks like we still want to keep original scores?

This is for rescoring and we want to overwrite the scores given the window. 

The idea here is for chaining rescorers together with various window sizes. I will add an integration test for chaining inference rescorer with a query rescorer to ensure things work as expected.